sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -0,0 +1,427 @@
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        super.populateDisplayData(builder);
+        builder.addIfNotNull(DisplayData.item("uri", uri));
+        builder.addIfNotNull(DisplayData.item("database", database));
+        builder.addIfNotNull(DisplayData.item("bucket", bucket));
+        builder.addIfNotNull(DisplayData.item("filterJson", filterJson));
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Coder<T> getDefaultOutputCoder() {
+        if (coder != null) {
+          return coder;
+        }
+        return (Coder<T>) SerializableCoder.of(Serializable.class);
+      }
+
+      class GridFSReader extends org.apache.beam.sdk.io.BoundedSource.BoundedReader<T> {
+        final BoundedGridFSSource<T> source;
+
+        Instant timestamp = Instant.now();
+        Iterator<ParseCallback.Line<T>> currentIterator;
+        ParseCallback.Line<T> currentLine;
+
+        GridFSReader(BoundedGridFSSource<T> source) {
+          this.source = source;
+        }
+
+        @Override
+        public boolean start() throws IOException {
+          setupGridFS();
+          if (objectIds == null) {
+            objectIds = new LinkedList<>();
+            DBCursor cursor = gridfs.getFileList().sort(null);
+            while (cursor.hasNext()) {
+              DBObject ob = cursor.next();
+              objectIds.add((ObjectId) ob.get("_id"));
+            }
+          }
+          return advance();
+        }
+
+        @Override
+        public boolean advance() throws IOException {
+          if (currentIterator != null && !currentIterator.hasNext()) {
+            objectIds.remove(0);
+            currentIterator = null;
+          }
+          if (currentIterator == null) {
+            if (objectIds.isEmpty()) {
+              return false;
+            }
+            ObjectId oid = objectIds.get(0);
+            GridFSDBFile file = gridfs.find(oid);
+            if (file == null) {
+              return false;
+            }
+            timestamp = new Instant(file.getUploadDate().getTime());
+            currentIterator = parser.parse(file);
+          }
+
+          if (currentIterator.hasNext()) {
+            currentLine = currentIterator.next();
+            return true;
+          }
+          return false;
+        }
+
+        @Override
+        public BoundedSource<T> getCurrentSource() {
+          return source;
+        }
+
+        @Override
+        public T getCurrent() throws NoSuchElementException {
+          if (currentLine != null) {
+            return currentLine.value;
+          }
+          throw new NoSuchElementException();
+        }
+
+        @Override
+        public Instant getCurrentTimestamp() throws NoSuchElementException {
+          if (currentLine != null) {
+            if (currentLine.timestamp != null) {
+              return currentLine.timestamp;
+            }
+            return timestamp;
```

Comment:
This should be "return Instant.now()" because otherwise the timestamp becomes completely arbitrary (time when the bundle started, which should be generally invisible to users - users don't care about bundle boundaries)